### PR TITLE
Refactorize Discretized Cruves and Landmarks

### PR DESCRIPTION
**Pull request by nguigs**
This addresses #377 and add the documentation. Also addressed #720

What about renaming discretized curves to discrete curves?

### CODE REVIEW

File: geomstats/geometry/discretized_curves.py [hunk 1]
```diff
@@ -15,25 +15,36 @@
...
 
     def belongs(self, point):
         """Test whether a point belongs to the manifold.
 
+        Test that all points on the curve belong to the manifold
+
         Parameters
         ----------
-        point : array-like
+        point : array-like, shape=[n_sampling_points, ambiant_dim]
```
Comment (ninamiolane) [inline geomstats/geometry/discretized_curves.py:40]: Typo ambient_dim

File: geomstats/geometry/landmarks.py [hunk 2]
```diff
@@ -1,229 +1,102 @@
...
+    """Class for space of landmarks.
+
+    The landmark space is a product manifold where all manifolds in the
+    product are the same. The default metric is the product metric and
+    is often referred to as the L2 metric.
+
+    Parameters
+    ----------
+    ambient_manifold : Manifold
+        Manifold to which landmarks belong
```
Comment (ninamiolane) [inline geomstats/geometry/landmarks.py:19]: Nit: missing end "."

File: geomstats/geometry/landmarks.py [hunk 3]
```diff
@@ -1,229 +1,102 @@
...
+    n_landmarks: int
+            Number of landmarks.
```
Comment (ninamiolane) [inline geomstats/geometry/landmarks.py:21]: NIT: space before : and indentation too large.

File: tests/test_poincare_polydisk.py [hunk 4]
```diff
@@ -46,7 +46,7 @@ def test_product_distance_extrinsic_representation(self):
 
         single_disk = PoincarePolydisk(n_disks=1, coords_type=coords_type)
         two_disks = PoincarePolydisk(n_disks=2, coords_type=coords_type)
-
+        print(point_b.shape)
```
Comment (ninamiolane) [inline tests/test_poincare_polydisk.py:49]: remaining print

File: geomstats/geometry/discretized_curves.py [hunk 5]
```diff
@@ -15,25 +15,36 @@
...
-    """Space of discretized curves sampled at points in ambient_manifold."""
+    """Space of discretized curves sampled at points in ambient_manifold.
+
+    Parameters
+    ----------
+    ambiant_manifold : Manifold
+        Manifold in which curves take values.
+    """
 
     def __init__(self, ambient_manifold):
```
Comment (ninamiolane) [inline geomstats/geometry/discretized_curves.py:26]: Why not having n_landmarks as a parameter of the constructor, like for the class `Landmarks`?
Or having `lambda n : L2Metric(manifold, n)` for both?

File: tests/test_discretized_curves.py [hunk 6]
```diff
@@ -256,7 +257,9 @@ def test_srv_metric_dist_and_geod(self):
         srv = self.srv_metric_r3.square_root_velocity(geod)
 
         srv_derivative = self.n_discretized_curves * (srv[1:, :] - srv[:-1, :])
-        norms = self.l2_metric_r3.norm(srv_derivative, geod[:-1, :-1, :])
+        l2_metric = self.space_curves_in_euclidean_3d.l2_metric(
+            self.n_sampling_points - 1)
```
Comment (ninamiolane) [inline tests/test_discretized_curves.py:261]: Why do we need `self.n_sampling_points -1` here, and in setup we have:
```
self.l2_metric_r3 = self.space_curves_in_euclidean_3d.l2_metric(
            self.n_sampling_points)
```

File: geomstats/geometry/discretized_curves.py [hunk 7]
```diff
@@ -207,23 +229,22 @@ def exp(self, tangent_vec, base_curve):
...
         coef_1 = 1 / gs.sqrt(base_curve_velocity_norm)
         coef_2 = -1 / (2 * base_curve_velocity_norm**(5 / 2)) * inner_prod
 
         term_1 = gs.einsum('ij,ijk->ijk', coef_1, tangent_vec_derivative)
         term_2 = gs.einsum('ij,ijk->ijk', coef_2, base_curve_velocity)
         srv_initial_derivative = term_1 + term_2
 
-        end_curve_srv = self.l2_metric.exp(tangent_vec=srv_initial_derivative,
-                                           base_landmarks=base_curve_srv)
+        end_curve_srv = self.l2_metric(n_sampling_points - 1).exp(
```
Comment (ninamiolane) [inline geomstats/geometry/discretized_curves.py:246]: Why `n_sampling_points - 1`?

File: geomstats/geometry/discretized_curves.py [hunk 8]
```diff
@@ -44,73 +55,83 @@ class SRVMetric(RiemannianMetric):
...
+        base_curve : array-like, shape=[..., n_sampling_points, ambiant_dim]
+            Point representing a discretized curve.
 
         Returns
         -------
-        norm :
+        norm : array-like, shape=[..., n_sampling_points]
+            Point-wise norms.
         """
         # TODO: Revise this to refer to action on single elements
```
Comment (ninamiolane) [inline geomstats/geometry/discretized_curves.py:131]: Do you understand this TODO? x)

File: geomstats/geometry/discretized_curves.py (hunk 1 above)
Comment (ninamiolane) [reply geomstats/geometry/discretized_curves.py:40]: (also at other places in the PR)

File: geomstats/geometry/landmarks.py (hunk 2 above)
Comment (ninamiolane) [reply geomstats/geometry/landmarks.py:19]: (also at other places in the PR)

File: geomstats/geometry/landmarks.py (hunk 3 above)
Comment (ninamiolane) [reply geomstats/geometry/landmarks.py:21]: (also at other places in the PR)

**Comment by nguigs** [comment]
> Great! Do we want to have different logics for Landmarks and DiscretizedCurves? Here, one takes n_landmarks as a parameter, and the other doesn't. Is this because, in applications, we may want to change the sampling along the curve, while the number of landmarks rarely changes?
> 
> Yes to change to DiscreteCurves if you think this makes more sense.
> 
> Why do we have sometimes `n_sampling_points` and then `n_sampling_points - 1` in the code?

Have a look at the discussion in #377, the whole point of having both Landmarks and DiscreteCurves classes is that they don't behave the same: in the landmark geometry literature the number of landmarks is usually fixed and landmarks are supposed to represent features, whereas in the discrete curve the number of sampling points is not, and results from a random process. That being said, I am not sure the current implementation of discrete curve works for different number of sampling points, but it is what we should be aiming for.

About `n_sampling_points` and `n_sampling_points - 1`, it results from the method `square_root_velocity` in the SRV metric and a different coefficient is applied to the initial point in e.g. distance, but I am not familiar with the theory, I just put the parameter so that the tests pass! Maybe @AliceLeBrigant can tell us more about it!

https://github.com/geomstats/geomstats/blob/c70315cca98ff604b2137a9cc676c1555d1910bd/geomstats/geometry/discretized_curves.py#L368-L373

File: geomstats/geometry/discretized_curves.py (hunk 8 above)
Comment (nguigs) [reply geomstats/geometry/discretized_curves.py:131]: No...

File: geomstats/geometry/discretized_curves.py (hunk 7 above)
Comment (nguigs) [reply geomstats/geometry/discretized_curves.py:246]: it was already the case
https://github.com/geomstats/geomstats/blob/932fe5a05da9d0f1622b51c152d3ab34d3e6ae3a/geomstats/geometry/discretized_curves.py#L230

File: tests/test_discretized_curves.py (hunk 6 above)
Comment (nguigs) [reply tests/test_discretized_curves.py:261]: in the setup it is used as the L2 metric, here it is used to test the SRV metric, which results in `n_sampling_points - 1` points

**Comment by nguigs** [comment]
The vectorization raises a NotImplementedError with Pytorch, and the SRV metric only works with numpy. Should this be mentionned in the doc?

**Comment by ninamiolane** [comment]
Thanks for the explanation on the different logic for DiscreteCurves and Landmarks.

Which not implemented error are you referring to for vectorization? OK to mention it in the doc or raise an issue.

File: geomstats/geometry/discretized_curves.py [hunk 9]
```diff
@@ -14,120 +14,172 @@
...
 
         Returns
         -------
         belongs : bool
+            Boolean evaluating if point belongs to the space of discrete
+            curves.
         """
-        belongs = gs.all(self.ambient_manifold.belongs(point))
-        return belongs
+        def each_belongs(pt):
```
[the remark this answers is not in the document]
Comment (nguigs) [reply geomstats/geometry/discretized_curves.py:65]: but this is a function within a method, and point is already used in the method

**Comment by nguigs** [comment]
> Which not implemented error are you referring to for vectorization? OK to mention it in the doc or raise an issue.

https://github.com/geomstats/geomstats/blob/c70315cca98ff604b2137a9cc676c1555d1910bd/geomstats/_backend/pytorch/__init__.py#L71

File: geomstats/geometry/discretized_curves.py [hunk 10]
```diff
@@ -14,120 +14,172 @@
...
+            if hasattr(ambient_manifold, 'metric'):
+                self.ambient_metric = ambient_manifold.metric
+            else:
+                raise ValueError('Either a metric must be passed or an '
+                                 'ambient manifold equipped with a metric.')
+        else:
+            self.ambient_metric = metric
+        self.l2_metric = lambda n: L2Metric(ambient_manifold, n_landmarks=n)
 
     def pointwise_inner_product(self, tangent_vec_a, tangent_vec_b,
```
[the remark this answers is not in the document]
Comment (nguigs) [reply geomstats/geometry/discretized_curves.py:109]: I had the feeling their was a difference in the output shape (in L2 it is summed over the dimension representing the sampling points), I'll check

**Pull request merged**